### ISSUE TITLE
how to ignore certain folders?

### ISSUE THREAD

**Issue by cwrau** [mvdan/gofumpt#304]
In gitlab CI we'd like to cache the go packages, via `GOPATH=$CI_PROJECT_DIR/.go`, but this tool then fails to run , as it tries to format `.go/pkg/mod/github.com/onsi/ginkgo/v2@v2.14.0/integration/_fixtures/reporting_fixture/malformed_sub_package/malformed_sub_package_suite_test.go:15:1: expected declaration, found NO_COMPILE`

Is there a way to ignore a folder? Or maybe generally `GOPATH`? Or stuff inside hidden directories?

**Comment by kolyshkin** [comment]
Seems like a duplicate of https://github.com/mvdan/gofumpt/issues/250

**Comment by kolyshkin** [comment]
@cwrau you can probably move `GOPATH` out of `$CWD` (or vise versa) to fix the issue you're having.

**Comment by cwrau** [comment]
> @cwrau you can probably move `GOPATH` out of `$CWD` (or vise versa) to fix the issue you're having.

Actually not, as I'm in CI in gitlab and can only cache folders inside the project directory, which is the git tree, which is the right where the main go.mod lies 😅

**Comment by kolyshkin** [comment]
@cwrau it's hard to figure out your setup with the info provided. Can you give a link to your ci setup/scripts?

**Comment by cwrau** [comment]
> @cwrau it's hard to figure out your setup with the info provided. Can you give a link to your ci setup/scripts?

It's nothing special, I'm checking out the repo, with `go.mod` at the root, set `GOPATH` to `$CI_PROJECT_DIR/.go` so I can cache it, and run `gofumpt`.

Currently I'm already workarounding this with `gofumpt -w api cmd pkg test`, but this is inflexible in case we add new directories.

I'd rather do the opposite and exclude `$GOPATH` instead of including everything else.

**Comment by kolyshkin** [comment]
I'm still missing some bits information here so I have to guess/assume/imagine a few things. Not seeing the full picture makes it harder to help, and might render my answers less useful (or just plain wrong). Surely, I can ask more questions about your setup details (which might be obvious to you but not me), but this is not the place to do so.

But let's go one step back.

>> @cwrau you can probably move GOPATH out of $CWD (or vise versa) to fix the issue you're having.

> Actually not, as I'm in CI in gitlab and can only cache folders inside the project directory, which is the git tree, which is the right where the main go.mod lies 😅

I believe you still can do that (move .go out of your source tree), and there are many ways to do so:
1. Use `GIT_CLONE_PATH` (https://docs.gitlab.com/ee/ci/runners/configure_runners.html#custom-build-directories).
2. Move the checked out sources to a different directory.
3. Use `GIT_STRATEGY:none` to disable automatic checkout, and run `git clone` manually.
...

**Comment by cwrau** [comment]
> I'm still missing some bits information here so I have to guess/assume/imagine a few things. Not seeing the full picture makes it harder to help, and might render my answers less useful (or just plain wrong). Surely, I can ask more questions about your setup details (which might be obvious to you but not me), but this is not the place to do so.
> 
> But let's go one step back.
> 
> > > @cwrau you can probably move GOPATH out of $CWD (or vise versa) to fix the issue you're having.
> 
> > Actually not, as I'm in CI in gitlab and can only cache folders inside the project directory, which is the git tree, which is the right where the main go.mod lies 😅
> 
> I believe you still can do that (move .go out of your source tree), and there are many ways to do so:
> 
> 1. Use `GIT_CLONE_PATH` ([docs.gitlab.com/ee/ci/runners/configure_runners.html](https://docs.gitlab.com/ee/ci/runners/configure_runners.html#custom-build-directories)).
> 2. Move the checked out sources to a different directory.
> 3. Use `GIT_STRATEGY:none` to disable automatic checkout, and run `git clone` manually.
>    ...

Of those three, number one would be the most likely one we'd be inclined to try, but I think we'd rather wait on #250, however I managed to miss that issue before creating this one 😅

But that does sound like it would work 👍

**Comment by mvdan** [comment]
Indeed this is a duplicate of #250, as @kolyshkin points out. See https://github.com/mvdan/gofumpt/issues/250#issuecomment-2247329624.
